src/rhsmlib/cloud/providers/aws.py

```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2020 Red Hat, Inc.
+#
+# This software is licensed to you under the GNU General Public License,
+# version 2 (GPLv2). There is NO WARRANTY for this software, express or
+# implied, including the implied warranties of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+# along with this software; if not, see
+# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+#
+# Red Hat trademarks are not licensed under GPLv2. No permission is
+# granted to use or replicate Red Hat trademarks that are incorporated
+# in this software or its documentation.
+#
+
+"""
+This is module implementing detector and metadata collector of virtual machine running on AWS
+"""
+
+from rhsmlib.cloud.detector import CloudDetector
+
+
+class AWSCloudDetector(CloudDetector):
+    """
+    Detector of cloud machine
+    """
+
+    ID = 'aws'
+
+    def __init__(self, hw_info):
+        """
+        Initialize instance of AWSCloudDetector
+        """
+        super(AWSCloudDetector, self).__init__(hw_info)
+
+    def is_vm(self):
+        """
+        Is system running on virtual machine or not
+        :return: True, when machine is running on VM; otherwise return False
+        """
+        return super(AWSCloudDetector, self).is_vm()
+
+    def is_running_on_cloud(self):
+        """
+        Try to guess if cloud provider is AWS using collected hardware information (output of dmidecode,
+        virt-what, etc.)
+        :return: True, when we detected sign of AWS in hardware information; Otherwise return False
+        """
+
+        # The system has to be VM
+        if self.is_vm() is False:
+            return False
+        # This is valid for AWS systems using Xen
+        if 'dmi.bios.version' in self.hw_info and 'amazon' in self.hw_info['dmi.bios.version']:
+            return True
+        # This is valid for AWS systems using KVM
+        if 'dmi.bios.vendor' in self.hw_info and 'Amazon EC2' in self.hw_info['dmi.bios.vendor']:
+            return True
+        # Try to get output from virt-what
+        if 'virt.host_type' in self.hw_info and 'aws' in self.hw_info['virt.host_type']:
+            return True
+        # In other cases return False
+        return False
+
+    def is_likely_running_on_cloud(self):
+        """
+        Return non-zero value, when the machine is virtual machine and it is running on kvm/xen and
+        some Amazon string can be found in output of dmidecode
+        :return: Float value representing probability that vm is running on AWS
+        """
+        probability = 0.0
+
+        # When the machine is not virtual machine, then there is probably zero chance that the machine
+        # is running on AWS
+        if self.is_vm() is False:
```

Comment:
We should account for the baremetal machines that exist on AWS if at all possible. The system being a vm definitely raises the likelihood that it is running in the cloud.

Reply:
I agree.

Reply:
Are we going to account for baremetal in this PR?
I don't want to block this PR on future work but we should handle baremetal before this feature is complete.

Reply:
I would skip support for bare metal instances in this PR. I'm not able to start bare metal instance. All attempts were terminated with some error (e.g. "The requested configuration is currently not supported.").

Reply:
Based on Jiri's investigation this as written already accounts for baremetal machines.